### PR TITLE
Tune image layout loop

### PR DESCRIPTION
**Pull request by Tony-LunarG**
Loop through intervals instead of subresources in an attempt
to improve performance, specifically of a trace of Counter-Strike

### CODE REVIEW

**Comment by mark-lunarg** [comment]
Was the original code wrong?  Can you share why this is better and what we give up -- I'm kinda lost.

**Comment by mark-lunarg** [comment]
@jzulauf-lunarg, can you provide some context por favor?

File: layers/buffer_validation.cpp [hunk 1]
```diff
@@ -621,7 +621,7 @@ bool CoreChecks::VerifyFramebufferAndRenderPassLayouts(RenderPassCreateVersion r
                 LayoutUseCheckAndMessage layout_check(subresource_map);
                 bool subres_skip = false;
                 auto pos = subresource_map->Find(view_state->normalized_subresource_range);
-                for (; pos != subresource_map->End() && !subres_skip; ++pos) {
```
Comment (jzulauf-lunarg) [inline layers/buffer_validation.cpp:624]: In the original, we walk through every subresource in the range, even if the state is the same as the previous subresource.

File: layers/buffer_validation.cpp [hunk 2]
```diff
@@ -621,7 +621,7 @@ bool CoreChecks::VerifyFramebufferAndRenderPassLayouts(RenderPassCreateVersion r
...
+                for (; !(pos.AtEnd()) && !subres_skip; pos.IncrementInterval()) {
```
Comment (jzulauf-lunarg) [inline layers/buffer_validation.cpp:624]: Here, we take advantage of the fact that behind the scenes the image layout is run-length compressed (stored in ranges of subresources over which the state (both initial layout and current layout) is constant.  We take the first subresource of the range as a proxy for the whole constant value range, and validate that against the requirement -- reporting on the first only, but that's the same as we used to skip the rest of the range when the first failed (subres_skip ... actually that logic could be simplified out at a later time)

The "IncrementInterval" skips over all the subresources that have the same state as we just checked, getting us to the next "constant value" range.

File: layers/image_layout_map.cpp [hunk 3]
```diff
@@ -312,6 +312,12 @@ void ImageSubresourceLayoutMap::ConstIterator::Increment() {
         pos_.subresource = range_gen_.GetSubresource();
     }
 }
+
+void ImageSubresourceLayoutMap::ConstIterator::IncrementInterval() {
+    current_index_ = constant_value_bound_;
```
Comment (jzulauf-lunarg) [inline layers/image_layout_map.cpp:317]: constant_value_bound_ is the exclusive upper bound of the constant value range.  When we force the current index to point to that, UpdateRangeAndValue skips to the next constant value range, setting that state as the current position/state for the iterator.

**Review by jzulauf-lunarg — COMMENTED** [review]
Added explanation of the what and why.  The whole thing that could be optionally done would be to cleanup the subres_skip vs. skip logic.  Subres skip was needed when we wanted to bail on continuing to report the same error over and over (because in the old version we were looking at the same state over and over)... but may not be needed.

@mark-lunarg "by your command"

**Review by jzulauf-lunarg — APPROVED** [review]
Good comments :)

**Pull request merged**